installer/generate_secrets.py

```diff
@@ -1,233 +1,297 @@
+            )
+        else:
+            raise Exception("Invalid auth provider")
+
+    def _pull_secret(self):
+        self.input_file(
+            "pull-secret", ".dockerconfigjson", ".docker/config.json to pull images"
+        )
 
-def generate_files(secrets):
-    print("Generating secrets files in secrets/")
-    os.makedirs("secrets", exist_ok=True)
+    def _ingress_nginx(self):
+        self.input_file("ingress-nginx", "tls.key", "Certificate private key")
+        self.input_file("ingress-nginx", "tls.crt", "Certificate chain")
 
-    for k, v in secrets.items():
-        with open(f"secrets/{k}", "w") as f:
-            f.write(json.dumps(v))
+    def _argocd(self):
+        self.input_field(
+            "installer", "argocd.admin.plaintext_password", "Admin password for ArgoCD?"
+        )
+        pw = self.secrets["installer"]["argocd.admin.plaintext_password"]
+        h = bcrypt.hashpw(pw.encode("ascii"), bcrypt.gensalt(rounds=15)).decode("ascii")
+        now_time = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
+
+        self._set_generated("argocd", "admin.password", h)
+        self._set_generated("argocd", "admin.passwordMtime", now_time)
+        self._set_generated("argocd", "server.secretkey", secrets.token_hex(16))
+
+    def _portal(self):
+        pw = secrets.token_hex(32)
+        self._set_generated("portal", "ADMIN_PASSWORD", pw)
+
+
+class OnePasswordSecretGenerator(SecretGenerator):
+    def __init__(self, environment, regenerate):
+        super().__init__(environment, regenerate)
+        self.op_secrets = {}
+        self.op = OnePassword()
+        self.parse_vault()
+
+    def parse_vault(self):
+        items = self.op.list_items("RSP-Vault")
+
+        for i in items:
+            key = None
+            environments = []
+            doc = self.op.get_item(uuid=i["uuid"])
+
+            for section in doc["details"]["sections"]:
+                for field in section["fields"]:
+                    if field["t"] == "generate_secrets_key":
+                        if key is None:
+                            key = field["v"]
+                        else:
+                            raise Exception("Found two generate_secrets_keys for {key}")
+                    elif field["t"] == "environment":
+                        environments.append(field["v"])
+
+            # The type of secret is either a note or a password login.
+            # First, check the notes.
+            secret_value = doc["details"]["notesPlain"]
+
+            # If we don't find anything, pull the password from a login item.
+            if not secret_value:
+                for f in doc["details"]["fields"]:
+                    if f["designation"] == "password":
+                        secret_value = f["value"]
+
+            if self.environment in environments:
+                self.op_secrets[key] = secret_value
+            elif not environments and self.op_secrets.get(key, None) == None:
```

Comment:
Minor nit: Prefer `is None` to `== None` in Python because some things that are not `None` can compare equal to it.

Reply:
Looking at this again, I think you just want the last part to be `and key not in self.op_secrets`.

Reply:
yeah, you're right, I'll do that